tijanabekic/assignment3/Assignment 3,new.java

```diff
@@ -0,0 +1,167 @@
+                return true;
+            }
+        }
+        return false;
+    }
+    
+    public void addWord(String word) {
+        words.add(word);
+        
+    }
+
+}
+
+
+public class A3Q1 {
+    public static Dictionary dictionary;
+    public static void checkWord(char[][] grid, int i, int j, String start, int m, int n, HashMap<String, Boolean> map, HashSet<String> set) {
+        
+        //check if out of bounds
+        if (i < 0 || j < 0 || i >= m || j >= n) {
+            return;
+        }
+        //make the string for checking
+        start =  start + grid[i][j];
+        
+        //add to map if the word is not there
+        if (!map.containsKey(start)) {
+            map.put(start, dictionary.isPrefix(start));
+        }
+        //if the string is not a prefix, it can never form a word
+        if (map.get(start) == false) {
+            return;
+        }
+        //add to result if it is a word
+        if (dictionary.isWord(start)) {
+            //by using a set we can eliminate duplicats
+            if (!set.contains(start)){
+                set.add(start);
+            }
+        }
+        //check if it is possible to make a word in any direction
+       
+        checkWord(grid, i - 1, j - 1, start, m, n, map, set);
+        checkWord(grid, i - 1, j, start, m, n, map, set);
+        checkWord(grid, i - 1, j + 1, start, m, n, map, set);
+        checkWord(grid, i, j - 1, start, m, n, map, set);
+        checkWord(grid, i, j + 1, start, m, n, map, set);
+        checkWord(grid, i + 1, j - 1, start, m, n, map, set);
+        checkWord(grid, i + 1, j, start, m, n, map, set);
+        checkWord(grid, i + 1, j + 1, start, m, n, map, set);
+
+        return;
+    }
+
+    public static List<String> findWords(char[][] grid) {
+        
+        if (grid == null) {
+            return null;
+        }
+        int m = grid.length;
+        int n = grid[0].length;
+        //store all the results in a list
+        LinkedList<String> result = new LinkedList<String>();
+
+        //make a map so there is no need to double check strings
+        HashMap<String, Boolean> map = new HashMap<String, Boolean>();
+        HashSet<String> set = new HashSet<String>();
+        //start at any point
+        for (int i = 0; i < m; i++) {
+            for (int j = 0; j < n; j++) {
+                String start = "";
+                checkWord(grid, i, j, start, m, n, map, set);
+            }
+        }
+        
+        Iterator iter = set.iterator();
+        while (iter.hasNext()) {
+            result.add((String) iter.next());
+        }
+        return result;
+    }
+
+    public static void testRegular() {
+        int m = 2;
+        int n = 3;
+        char[][] grid = new char[m][n];
+        grid[0][0] = 'A'; grid[0][1] = 'A'; grid[0][2] = 'R';
+        grid[1][0] = 'T'; grid[1][1] = 'C'; grid[1][2] = 'D';
+        List<String> result = findWords(grid);
+        if (result.size() == 3) {

```

Comment:
If the dictionary contains DAD then I think that this test will fail. You are not checking for re-use of individual letters.